### PR TITLE
Performance optimizations for RandomStringUtils

### PR DESCRIPTION
**Pull request by fabrice102**
This commit improves the performance of RandomStringUtils:

* Reduces the number of random bytes generated and the number of calls to the random number generator, by using a cache system `AmortizedRandomBits`.
* Optimizes the case of alphanumerical strings, reducing the number of rejections in the rejection sampling.

See comments in code for details.

### CODE REVIEW

File: src/main/java/org/apache/commons/lang3/AmortizedRandomBits.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * AmortizedRandomBits enable to generate random integers of specific bit length.
+ *
+ * <p>It is more efficient way than calling Random.nextInt(1 << nbBits). It uses a cache of
+ * cacheSize random bytes that it replenishes when it gets empty. This is especially beneficial for
+ * SecureRandom Drbg implementations that incur a constant cost at each randomness generation. It is
+ * not thread safe.
+ *
+ * <p>Used internally by RandomStringUtils.
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/lang3/AmortizedRandomBits.java]: Please close HTML tags.

File: src/main/java/org/apache/commons/lang3/AmortizedRandomBits.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ */
+class AmortizedRandomBits {
+    private final Random random;
+
+    private final byte[] cache;
+
+    // bitIndex is the index of the next bit in the cache to be used
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/lang3/AmortizedRandomBits.java]: Use the Javadoc format.

**Review by garydgregory — CHANGES_REQUESTED** [review]
Hello @fabrice102 !
Thank you for your contribution.
I'm OK with the PR, but, I'd like to resolve the discussion around the change in behavior this PR addresses first before making changes geared toward performance claims.
I've made some low-level comments.
Also, note the build failures due to checkstyle issues.
In the future, run `mvn` by itself to run all build checks before pushing.
TY! 😄

**Comment by fabrice102** [comment]
Hi @garydgregory! Thanks for the feedback! I'll submit a new version on Monday.

**Pull request merged**